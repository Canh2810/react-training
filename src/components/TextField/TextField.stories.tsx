import { StoryObj, Meta } from '@storybook/react'

import TextField from '.'
import { TextFieldVariants } from '@types'
import { Search } from '@assets/icons'

export default {
  title: 'Components/TextField',
  component: TextField,
} as Meta

type Story = StoryObj<typeof TextField>

export const Default: Story = {
  args: {
    icon: <Search />,
    placeholder: 'Text Field',
    onChange: () => {},
  },
}

export const Filled: Story = {
  args: {
    variant: TextFieldVariants.Filled,
    icon: <Search />,
    placeholder: 'Text Field',
    onChange: () => {},
  },
}

export const HasErrorMessage: Story = {
  args: {
    icon: <Search />,
    placeholder: 'Text Field',
    errorMessage: 'Field is required',
    onChange: () => {},
  },
}
