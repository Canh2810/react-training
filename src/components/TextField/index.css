.text-field {
  position: relative;
  display: flex;
  align-items: center;
  padding: 12px 16px;
  gap: 10px;
  border-radius: 6px;
  font-size: 16px;
}

.text-field input {
  border: none;
  width: calc(100% - 14px);
  background-color: unset;
  color: #a1a1aa;
}

.text-field input:focus {
  outline: none;
}

.text-field__error-message {
  position: absolute;
  top: 48px;
  font-size: var(--xs-font);
  color: var(--danger-color);
}
