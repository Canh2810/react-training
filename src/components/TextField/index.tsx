// Libs
import { ChangeEvent, memo } from 'react'

// Types
import { TextFieldTypes, TextFieldVariants } from '@types'

// Themes
import { variants } from '@themes'

// Stores
import { useThemeStore } from '@stores'

// Style
import './index.css'

export interface TextFieldProps {
  variant?: TextFieldVariants
  icon: JSX.Element
  name: string
  placeholder: string
  id: string
  value: string
  type?: TextFieldTypes
  errorMessage?: string
  className?: string
  onChange: (e: ChangeEvent<HTMLInputElement>) => void
}

const TextField = ({
  variant = TextFieldVariants.Outlined,
  icon,
  name,
  placeholder,
  id,
  value,
  type = TextFieldTypes.Text,
  className = '',
  errorMessage,
  onChange,
}: TextFieldProps) => {
  const { theme } = useThemeStore((state) => state)

  const textFieldStyle = variants[variant](theme)

  return (
    <div className={`text-field ${className}`} style={{ ...textFieldStyle }}>
      <input
        type={type}
        id={id}
        name={name}
        value={value}
        placeholder={placeholder}
        onChange={onChange}
      />
      {icon}
      {errorMessage && (
        <p className="text-field__error-message">{errorMessage}</p>
      )}
    </div>
  )
}

export default memo(TextField)
