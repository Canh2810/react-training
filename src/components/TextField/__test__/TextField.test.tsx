// PostList.test.js (or PostList.test.tsx for TypeScript)

import { fireEvent, render } from '@testing-library/react'
import { Search } from '@assets'
import TextField from '..'
import { TextFieldTypes, TextFieldVariants } from '@types'

describe('PostList component', () => {
  const mockChange = jest.fn()
  const props = {
    icon: <Search />,
    name: 'test',
    placeholder: 'test',
    id: 'test',
    value: 'test',
    onChange: mockChange,
  }

  it('render with variant is outlined and type is text', () => {
    const { container } = render(<TextField {...props} />)
    expect(container).toMatchSnapshot()
  })

  it('render with variant is filled and type is text', () => {
    const { container } = render(
      <TextField {...props} variant={TextFieldVariants.Filled} />,
    )
    expect(container).toMatchSnapshot()
  })

  it('render with variant is outlined and type is password', () => {
    const { container } = render(
      <TextField {...props} type={TextFieldTypes.Password} />,
    )
    expect(container).toMatchSnapshot()
  })
  it('render with variant is filled and type is password', () => {
    const { container } = render(
      <TextField
        {...props}
        variant={TextFieldVariants.Filled}
        type={TextFieldTypes.Password}
      />,
    )
    expect(container).toMatchSnapshot()
  })

  it('should call onChange handler when value changes', () => {
    const { getByPlaceholderText } = render(<TextField {...props} />)
    const textfield = getByPlaceholderText('test')

    fireEvent.change(textfield, { target: { value: 'abc' } })
    expect(mockChange).toHaveBeenCalled()
  })
})
