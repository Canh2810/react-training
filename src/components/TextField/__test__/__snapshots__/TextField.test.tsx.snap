// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostList component render with variant is filled and type is password 1`] = `
<div>
  <div
    class="text-field "
    style="background-color: rgb(244, 244, 245); color: rgb(161, 161, 170);"
  >
    <input
      id="test"
      name="test"
      placeholder="test"
      type="password"
      value="test"
    />
    <svg
      fill="none"
      height="16"
      width="16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M6.91 2a4.909 4.909 0 1 0-.001 9.818 4.909 4.909 0 0 0 0-9.818Z"
        stroke="#52525B"
        stroke-miterlimit="10"
        stroke-width="1.5"
      />
      <path
        d="M10.572 10.572 14 14"
        stroke="#52525B"
        stroke-linecap="round"
        stroke-miterlimit="10"
        stroke-width="1.5"
      />
    </svg>
  </div>
</div>
`;

exports[`PostList component render with variant is filled and type is text 1`] = `
<div>
  <div
    class="text-field "
    style="background-color: rgb(244, 244, 245); color: rgb(161, 161, 170);"
  >
    <input
      id="test"
      name="test"
      placeholder="test"
      type="text"
      value="test"
    />
    <svg
      fill="none"
      height="16"
      width="16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M6.91 2a4.909 4.909 0 1 0-.001 9.818 4.909 4.909 0 0 0 0-9.818Z"
        stroke="#52525B"
        stroke-miterlimit="10"
        stroke-width="1.5"
      />
      <path
        d="M10.572 10.572 14 14"
        stroke="#52525B"
        stroke-linecap="round"
        stroke-miterlimit="10"
        stroke-width="1.5"
      />
    </svg>
  </div>
</div>
`;

exports[`PostList component render with variant is outlined and type is password 1`] = `
<div>
  <div
    class="text-field "
    style="border: 1px solid; color: rgb(105, 106, 117); border-color: rgba(105, 106, 117, 0.30); background-color: rgb(255, 255, 255);"
  >
    <input
      id="test"
      name="test"
      placeholder="test"
      type="password"
      value="test"
    />
    <svg
      fill="none"
      height="16"
      width="16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M6.91 2a4.909 4.909 0 1 0-.001 9.818 4.909 4.909 0 0 0 0-9.818Z"
        stroke="#52525B"
        stroke-miterlimit="10"
        stroke-width="1.5"
      />
      <path
        d="M10.572 10.572 14 14"
        stroke="#52525B"
        stroke-linecap="round"
        stroke-miterlimit="10"
        stroke-width="1.5"
      />
    </svg>
  </div>
</div>
`;

exports[`PostList component render with variant is outlined and type is text 1`] = `
<div>
  <div
    class="text-field "
    style="border: 1px solid; color: rgb(105, 106, 117); border-color: rgba(105, 106, 117, 0.30); background-color: rgb(255, 255, 255);"
  >
    <input
      id="test"
      name="test"
      placeholder="test"
      type="text"
      value="test"
    />
    <svg
      fill="none"
      height="16"
      width="16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M6.91 2a4.909 4.909 0 1 0-.001 9.818 4.909 4.909 0 0 0 0-9.818Z"
        stroke="#52525B"
        stroke-miterlimit="10"
        stroke-width="1.5"
      />
      <path
        d="M10.572 10.572 14 14"
        stroke="#52525B"
        stroke-linecap="round"
        stroke-miterlimit="10"
        stroke-width="1.5"
      />
    </svg>
  </div>
</div>
`;
